### PR TITLE
Minor refactor in FiberCacheManger to support cache eviction

### LINKED ISSUE #239: Memory should assign a proper data capacity to avoid reach physical memory limit
Container killed by YARN for exceeding memory limits. 52.6 GB of 50 GB physical memory used. Consider boosting spark.yarn.executor.memoryOverhead

The reason is FiberCacheManager never evict cache.

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

1. Use configuration to assign a proper cache size
2. Use KB instead of B as Unit in `weigher` due to Guava known issue

## How was this patch tested?

Unit Test

Fixes #239

### CODE REVIEW

**Comment by linhong-intel** [comment]
@adrian-wang Need some suggestion from you.

File: src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala [hunk 1]
```diff
@@ -19,18 +19,18 @@ package org.apache.spark.sql.execution.datasources.spinach.filecache
 
 import java.util.concurrent.TimeUnit
 
-import scala.collection.mutable
-
-import collection.JavaConverters._
 import com.google.common.cache._
 import org.apache.hadoop.conf.Configuration
+import scala.collection.JavaConverters._
+import scala.collection.mutable
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: Nit: Import order should be java -> scala -> 3rd party -> local

File: src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala [hunk 2]
```diff
@@ -52,33 +52,63 @@ private[spinach] sealed case class ConfigurationCache[T](key: T, conf: Configura
...
-      .newBuilder()
+  @transient protected var cache: LoadingCache[ENTRY, FiberCache] = _
+
+  private def buildCache(conf: Configuration): LoadingCache[ENTRY, FiberCache] = {
+
+    val weight = conf.getLong(SQLConf.SPINACH_FIBERCACHE_SIZE.key,
+                              SQLConf.SPINACH_FIBERCACHE_SIZE.defaultValue.get)
+
+    if (weight >= 8 * 1024 * 1024 * 1024L) {
+      throw new SpinachException(s"${SQLConf.SPINACH_FIBERCACHE_SIZE.key} should be less than 8GB")
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: We'd better not throw exception, but put something in log and cap it to 8G

File: src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala [hunk 3]
```diff
@@ -52,33 +52,63 @@ private[spinach] sealed case class ConfigurationCache[T](key: T, conf: Configura
...
+    }
+
+    val builder = CacheBuilder.newBuilder()
       .concurrencyLevel(4) // DEFAULT_CONCURRENCY_LEVEL TODO verify that if it works
       .weigher(new Weigher[ENTRY, FiberCache] {
-        override def weigh(key: ENTRY, value: FiberCache): Int = value.fiberData.size().toInt
+        override def weigh(key: ENTRY, value: FiberCache): Int = {
+          // Use KB as Unit due to large weight will cause Guava overflow
+          val weight = value.fiberData.size() / 1024
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: maybe ceil the result to ensure not overflow.

File: src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala [hunk 4]
```diff
@@ -52,33 +52,63 @@ private[spinach] sealed case class ConfigurationCache[T](key: T, conf: Configura
...
+            throw new SpinachException(s"Too large fiber size $weight KB")
+          }
+          weight.toInt
+        }
       })
-      .maximumWeight(MemoryManager.getDataCacheCapacity())
+      .maximumWeight(weight)
       .removalListener(new RemovalListener[ENTRY, FiberCache] {
         override def onRemoval(n: RemovalNotification[ENTRY, FiberCache]): Unit = {
+          logDebug("Loading Fiber Cache" + (n.getKey.key match {
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: This is removing

File: src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala (hunk 3 above)
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: fixed

Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/spinach/filecache/FiberCacheManager.scala]: I mean, 2040 / 1024 = 1, but maybe when take it as 2 is better.

**Comment by linhong-intel** [comment]
I'm planning to merge it by today.

**Pull request merged**